tfhe/src/core_crypto/commons/mod.rs

```diff
@@ -77,24 +71,50 @@ pub mod test_tools {
     use crate::core_crypto::commons::traits::*;
     use concrete_csprng::seeders::{Seed, Seeder};
 
-    fn modular_distance<T: UnsignedInteger>(first: T, other: T) -> T {
+    pub fn modular_distance<T: UnsignedInteger>(first: T, other: T) -> T {
         let d0 = first.wrapping_sub(other);
         let d1 = other.wrapping_sub(first);
         d0.min(d1)
     }
 
-    fn torus_modular_distance<T: UnsignedInteger>(first: T, other: T) -> f64 {
-        let d0 = first.wrapping_sub(other);
-        let d1 = other.wrapping_sub(first);
-        if d0 < d1 {
-            let d: f64 = d0.cast_into();
-            d / 2_f64.powi(T::BITS as i32)
+    pub fn torus_modular_distance<T: UnsignedInteger>(
+        first: T,
+        other: T,
+        modulus: CiphertextModulus<T>,
+    ) -> f64 {
+        if modulus.is_compatible_with_native_modulus() {
+            if modulus.is_native_modulus() {
+                let d0 = first.wrapping_sub(other);
+                let d1 = other.wrapping_sub(first);
+                if d0 < d1 {
+                    let d: f64 = d0.cast_into();
+                    d / 2_f64.powi(T::BITS as i32)
+                } else {
+                    let d: f64 = d1.cast_into();
+                    -d / 2_f64.powi(T::BITS as i32)
+                }
+            } else {
+                let modulus = modulus.get_custom_modulus();
+                let d0 = first.wrapping_sub(other);
+                let d1 = other.wrapping_sub(first);
+                if d0 < d1 {
+                    let d: f64 = d0.cast_into();
+                    d / 2_f64.powi(modulus.ilog2() as i32)
+                } else {
+                    let d: f64 = d1.cast_into();
+                    -d / 2_f64.powi(modulus.ilog2() as i32)
+                }
+            }
         } else {
-            let d: f64 = d1.cast_into();
-            -d / 2_f64.powi(T::BITS as i32)
+            todo!("Currently unimplemented for non power of 2 moduli")
         }
     }
 
+    pub fn variance(samples: &[f64]) -> Variance {
```

Comment:
Should be add an assert! / debug_assert! on (num_sample > 1) ?
otherwise we may have division by zero / overflow into division by usize::MAX

Reply:
apparently it does yes